core/iwasm/common/wasm_application.c

```diff
@@ -107,7 +107,22 @@ execute_main(WASMModuleInstanceCommon *module_inst, int32 argc, char *argv[])
        the actual main function. Directly calling main function
        may cause exception thrown. */
     if ((func = wasm_runtime_lookup_wasi_start_function(module_inst))) {
-        return wasm_runtime_call_wasm(exec_env, func, 0, NULL);
+        const char *wasi_proc_exit_exception = "wasi proc exit";
+
+        ret = wasm_runtime_call_wasm(exec_env, func, 0, NULL);
+        /* a successful return from _start is same as proc_exit(0). */
+        if (ret) {
+            wasm_runtime_set_exception(module_inst, wasi_proc_exit_exception);
+            /* exit_code is zero-initialized */
+            ret = false;
```

Comment:
Not sure why need to set `wasi_proc_exit_exception` here? If ret is true, there should be no exception thrown and wasi_exit_code was already initialized as 0, seems we can do nothing?

Reply:
But the main thread will wait until all other threads exit in wasm_exec_env_destroy, it doesn't meet the "wasi proc exit" exception, seems no need to propagate "wasi proc exit" for other threads to exit early? Or other threads may not finish their jobs normally, also other threads may fail to throw exception if they continue to run?

Reply:
That's what I my concern is, suppose the sub thread is doing an important job, e.g. do some calculation, output important data to a file, can we terminate it from the main thread? Is it better to wait until the sub threads exit? In fact, there is API wasm_cluster_wait_for_all_except_self in thread manager.

Not sure whether there is reference about "a return from _start should behave as exit(0)" in wasi-libc?

@loganek What's your opinion?

Reply:
i'm not sure if i understand your concern.
the importance of the job is something apps should consider. not runtime.

> Not sure whether there is reference about "a return from _start should behave as exit(0)" in wasi-libc?

* a return from main() is an equivalent of exit(). (C standard)
* wasi-libc maps a return from main() to a return from `_start` when the exit code is 0. https://github.com/WebAssembly/wasi-libc/blob/ec4566beae84e54952637f0bf61bee4b4cacc087/libc-bottom-half/crt/crt1-command.c#L48-L52

Reply:
I am not very sure whether it is a good behavior that the main thread terminates all others thread instead of waiting for them. But if you want to do that, can we just call the API `wasm_cluster_terminate_all_except_self` of thread manager:
```C
#if WASM_ENABLE_THREAD_MGR != 0
    if (ret) {
        WASMCluster *cluster = wasm_exec_env_get_cluster(exec_env);
        wasm_cluster_terminate_all_except_self(cluster, exec_env);
    }
#endif
```
It also sets the terminate flag for other threads.

Reply:
i chose to use wasm_runtime_set_exception because:

* what i want to do here is to mimic wasi proc_exit, which uses wasm_runtime_set_exception.
* wasm_cluster_terminate_all_except_self seems unused (thus untested) and broken. (cluster->lock vs cluster_list_lock locking order issue.)
* we don't need to join threads here

Reply:
OK, if to use wasm_runtime_set_exception, could you please add comments to mention somewhat like we want to mimic wasi proc_exit and to terminate other threads? It is not very easy to understand the code :)
And could you wrap the code with `#if WASM_ENABLE_THREAD_MGR != 0 .. #endif`, it is only for multi-threading?

Reply:
> OK, if to use wasm_runtime_set_exception, could you please add comments to mention somewhat like we want to mimic wasi proc_exit and to terminate other threads? It is not very easy to understand the code :) And could you wrap the code with `#if WASM_ENABLE_THREAD_MGR != 0 .. #endif`, it is only for multi-threading?

i prefer less #ifdef in general. but ok. done.